javascript/house-style.md

```diff
@@ -939,6 +939,33 @@ exports default { init };
 
 ### DOM binding
 
+When it comes time to attach your [JavaScript module](#module-architecture) to a specific DOM element (or collection of elements), we favour the use of descriptive `data-component` attributes in the markup rather than using specific class names or ids.
+
+We do this:
+
+```html
+<nav class="c-popout-nav c-popout-nav--dark" data-component="popout">
+  <button class="c-popout-nav__close" data-component="popout-close">Close</button>
+  (etc.)
+</nav>
+```
+
+We don't do this:
+
+```html
+<nav class="c-popout-nav c-popout-nav--dark" data-popout>...</nav>
+
+<nav class="js-popout c-popout-nav c-popout-nav--dark">...</nav>
+
+<nav id="popout" class="c-popout-nav c-popout-nav--dark">...</nav>
+```
+
+As demonstrated above, we use `data-component` attributes to label the "main" element that is being bound by the JavaScript, and also any additional elements used by the module which can't otherwise be unabiguously selected.  In instaces where child elements can be inferred without using additional attributes &mdash; for example a module which is expected to contain just a single input &mdash; then it may be acceptable to use other selectors such as type selectors, however care should be taken.
```

Comment:
"also any additional elements used by the module which can't otherwise be unabiguously selected"

On SpringerLink we use `data-role="name"` for these. The module is always bound to a `data-component` and any child can be given a `data-role`

Reply:
Ah interesting.  Yeah I thought this might be a contentious bit — it's something @davidpauljunior and I have talked about whether to reuse `data-component` or something else.  David had suggested another option was to re-use the component name e.g. `data-component="popout"` then children like `data-popout="close"`.

Personally I favour have more standardised attribute names, but using a single alternate one like `data-role` may be a middle-ground solution.

Reply:
Do we need to write about main and additional elements? I'm wondering if that's irrelevant really and maybe certain modules won't have a main hook but will have multiple hooks. Personally I like the idea that every js hook follows the same `data-component` convention, perhaps with a prefix to associate it with a js module, e.g. `data-component="popout"`, `data-component="popout-close"`, `data-component="popout-totally-amazing-parallax-carousel-figit-spinner"`

Reply:
That's a good point that the idea of a "main" element may not make much sense, depending on what the module is doing.

If we went this way I could change that paragraph to;

>As demonstrated above, we use `data-component` attributes to label any elements that are being bound by the JavaScript.  In instances where other related elements can be inferred without using additional data attributes &mdash; for example a module which is expected to contain just a single input &mdash; then it may be acceptable to use other selectors such as type selectors, however care should be taken.

Makes it quite a lot simpler.

Reply:
Agreed, this does simplify things and allows for more complex use cases 👍

Reply:
Labelling things as `data-component="whatever" data-component="another"` and then passing them into a module as needed looks great. The naming convention is flexible and allows the naming itself to imply child-parent relationships (e.g. `data-component="whatever"` and 'data-component="whatever-child"').

Like @benjclark says, there's going to be plenty of cases where you don't have a "main" component, so the removal of main makes sense to me.